specs/dotsigstore_bundle/pb/bundle.proto

```diff
@@ -0,0 +1,84 @@
+syntax = "proto3";
+package sigstore;
+
+// https://raw.githubusercontent.com/secure-systems-lab/dsse/9c813476bd36de70a5738c72e784f123ecea16af/envelope.proto
+import "envelope.proto";
+
+option go_package = "github.com/sigstore/xxx/bundle";
+
+enum HashAlgorithm {
+        SHA2_224 = 0;
+        SHA2_256 = 1;
+        SHA2_384 = 2;
+        SHA2_512 = 3;
+        SHA2_512_224 = 4;
+        SHA2_512_256 = 5;
+}
+
+message AttestationBlob {
+        // In JSON encoding this shall be base64 encoded.
+        bytes payload_hash = 1;
+        HashAlgorithm algorithm = 2;
+        // In JSON encoding this shall be base64 encoded.
+        bytes signature = 3;
+}
+
+// RekorEntry is a separate type, capturing all the details required
+// from Rekor to reconstruct an entry, given that the payload is provided
+// via other means.
+// This type can easily be created from the existing response from Rekor.
+// Future iterations could rely on Rekor returning the minimal set of
+// attributes (excluding the payload) that are required for verifying the
+// SET.
+message RekorEntry {
+        int64 log_index = 1;
+        string log_id = 2;
+        // Even if kind can be derived from the attestation type it's kept
+        // as this struct could be automatically created by the sigstore
+        // library based on the response from Rekor. Also during verification
+        // this struct can be passed to the library for Rekor entry
+        // reconstruction. By duplicating the kind here we would so minimize
+        // any possibility of error due to attestation/kind matching logic,
+        // as that logic must be hard-coded in the client. With this construct
+        // it's only determined during publish to Rekor.
+        string kind = 3;
+        string version = 4;
+        string signed_entry_timestamp = 5;
+        // Note that this will be encoded as a string in JSON:
+        // https://developers.google.com/protocol-buffers/docs/proto3#json
+        int64 integrated_time = 6;
+}
+
+message PublicKey {
+        // Optional unauthenticated hint on which key to use.
+        // If a DSSE envelope is signed, at least one of the signatures
+        // MUST have a signature with the same key id.
+        string key_id = 1;
+}
+
+message X509Cert {
+        // PEM encoded certificate.
+        string certificate = 1;
+        // PEM encoded certificate chain.
+        string chain = 2;
```

Comment:
Just a point of clarification (it wasn't obvious from the examples) . . . how should the certificate be stored

Base64-encoded PEM:
```
LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSUNvRENDQWlhZ0F3SUJBZ0lVTUVYRVVtR2NtWGhrNExrd1NXUnlzOEJyYWxzd0NnWUlLb1pJemowRUF3TXcKTnpFVk1CTUdBMVVFQ2hNTWMybG5jM1J2Y21VdVpHVjJNUjR3SEFZRFZRUURFeFZ6YVdkemRHOXlaUzFwYm5SbApjbTFsWkdsaGRHVXdIaGNOTWpJd09ESTFNVGt5TlRNeFdoY05Nakl3T0RJMU1Ua3pOVE14V2pBQU1Ga3dFd1lICktvWkl6ajBDQVFZSUtvWkl6ajBEQVFjRFFnQUVxTUFQYklCNTl4cnBRZURYcldhVklMTWR2RUI0ckhZVVd1ZjUKZklvVFZSSnRicWc0b1FMbFZHRVZGWGgrMytGd3l4cFhOaU9sV0ZTQ0tGdTNxZmRXMHFPQ0FVVXdnZ0ZCTUE0RwpBMVVkRHdFQi93UUVBd0lIZ0RBVEJnTlZIU1VFRERBS0JnZ3JCZ0VGQlFjREF6QWRCZ05WSFE0RUZnUVVTcUlMCkhsWjg1Nm9GY3pIMjJuN2llZE0zazZFd0h3WURWUjBqQkJnd0ZvQVUzOVBwejFZa0VaYjVxTmpwS0ZXaXhpNFkKWkQ4d0h3WURWUjBSQVFIL0JCVXdFNEVSWW5KcFlXNUFaR1ZvWVcxbGNpNWpiMjB3TEFZS0t3WUJCQUdEdnpBQgpBUVFlYUhSMGNITTZMeTluYVhSb2RXSXVZMjl0TDJ4dloybHVMMjloZFhSb01JR0tCZ29yQmdFRUFkWjVBZ1FDCkJId0VlZ0I0QUhZQUNHQ1M4Q2hTLzJoRjBkRnJKNFNjUldjWXJCWTl3empTYmVhOElnWTJiM0lBQUFHQzFuZCsKa0FBQUJBTUFSekJGQWlBa1h4VHU4MUY1VDA3Wm9UeUhTU1ZRcjI3NzRmREN3cHV0cnpQdjhzRzR3QUloQU5qWgpUYWlNSk01emFsdGhvcnZLSWNUNFpHYmsrdDBuV0tZUitBYjNDVzNaTUFvR0NDcUdTTTQ5QkFNREEyZ0FNR1VDCk1DbVZXK1RCWStvSzVNYkpnSGwrY3hJSzJ3L01XalAvcVFVaC95R2dLc0I0aVBxdXNUbENTSlJrcjF6QWVjQ2YKV0FJeEFMQXI0MnpoZTVXZzZDSlhobEVHVjZHWTlmN0YyNmxNVGJRdUVpTlJEazJuS1RSaU85MHlWNnZreGFjSgpRckV1N0E9PQotLS0tLUVORCBDRVJUSUZJQ0FURS0tLS0t
```

PEM:
```
-----BEGIN CERTIFICATE-----
MIICoDCCAiagAwIBAgIUMEXEUmGcmXhk4LkwSWRys8BralswCgYIKoZIzj0EAwMw
NzEVMBMGA1UEChMMc2lnc3RvcmUuZGV2MR4wHAYDVQQDExVzaWdzdG9yZS1pbnRl
cm1lZGlhdGUwHhcNMjIwODI1MTkyNTMxWhcNMjIwODI1MTkzNTMxWjAAMFkwEwYH
KoZIzj0CAQYIKoZIzj0DAQcDQgAEqMAPbIB59xrpQeDXrWaVILMdvEB4rHYUWuf5
fIoTVRJtbqg4oQLlVGEVFXh+3+FwyxpXNiOlWFSCKFu3qfdW0qOCAUUwggFBMA4G
A1UdDwEB/wQEAwIHgDATBgNVHSUEDDAKBggrBgEFBQcDAzAdBgNVHQ4EFgQUSqIL
HlZ856oFczH22n7iedM3k6EwHwYDVR0jBBgwFoAU39Ppz1YkEZb5qNjpKFWixi4Y
ZD8wHwYDVR0RAQH/BBUwE4ERYnJpYW5AZGVoYW1lci5jb20wLAYKKwYBBAGDvzAB
AQQeaHR0cHM6Ly9naXRodWIuY29tL2xvZ2luL29hdXRoMIGKBgorBgEEAdZ5AgQC
BHwEegB4AHYACGCS8ChS/2hF0dFrJ4ScRWcYrBY9wzjSbea8IgY2b3IAAAGC1nd+
kAAABAMARzBFAiAkXxTu81F5T07ZoTyHSSVQr2774fDCwputrzPv8sG4wAIhANjZ
TaiMJM5zalthorvKIcT4ZGbk+t0nWKYR+Ab3CW3ZMAoGCCqGSM49BAMDA2gAMGUC
MCmVW+TBY+oK5MbJgHl+cxIK2w/MWjP/qQUh/yGgKsB4iPqusTlCSJRkr1zAecCf
WAIxALAr42zhe5Wg6CJXhlEGV6GY9f7F26lMTbQuEiNRDk2nKTRiO90yV6vkxacJ
QrEu7A==
-----END CERTIFICATE-----
```

Base64-encoded DER (PEM w/o headers or line-breaks):
```
MIICoDCCAiagAwIBAgIUMEXEUmGcmXhk4LkwSWRys8BralswCgYIKoZIzj0EAwMwNzEVMBMGA1UEChMMc2lnc3RvcmUuZGV2MR4wHAYDVQQDExVzaWdzdG9yZS1pbnRlcm1lZGlhdGUwHhcNMjIwODI1MTkyNTMxWhcNMjIwODI1MTkzNTMxWjAAMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEqMAPbIB59xrpQeDXrWaVILMdvEB4rHYUWuf5fIoTVRJtbqg4oQLlVGEVFXh+3+FwyxpXNiOlWFSCKFu3qfdW0qOCAUUwggFBMA4GA1UdDwEB/wQEAwIHgDATBgNVHSUEDDAKBggrBgEFBQcDAzAdBgNVHQ4EFgQUSqILHlZ856oFczH22n7iedM3k6EwHwYDVR0jBBgwFoAU39Ppz1YkEZb5qNjpKFWixi4YZD8wHwYDVR0RAQH/BBUwE4ERYnJpYW5AZGVoYW1lci5jb20wLAYKKwYBBAGDvzABAQQeaHR0cHM6Ly9naXRodWIuY29tL2xvZ2luL29hdXRoMIGKBgorBgEEAdZ5AgQCBHwEegB4AHYACGCS8ChS/2hF0dFrJ4ScRWcYrBY9wzjSbea8IgY2b3IAAAGC1nd+kAAABAMARzBFAiAkXxTu81F5T07ZoTyHSSVQr2774fDCwputrzPv8sG4wAIhANjZTaiMJM5zalthorvKIcT4ZGbk+t0nWKYR+Ab3CW3ZMAoGCCqGSM49BAMDA2gAMGUCMCmVW+TBY+oK5MbJgHl+cxIK2w/MWjP/qQUh/yGgKsB4iPqusTlCSJRkr1zAecCfWAIxALAr42zhe5Wg6CJXhlEGV6GY9f7F26lMTbQuEiNRDk2nKTRiO90yV6vkxacJQrEu7A==
```

Given that the field potentially holds multiple certificates, I'm gonna guess it's either option 1 or 2 above.

Reply:
What if `chain` was configured as `bytes` rather than `string`?

Reply:
I would suggest using a PEM-encoded string for each certificate. Other approaches have caused issues in the past:
* Base64 encoding the PEM encoding - This is unnecessary, it's just a double encoding
* Base64-encoded DER without the headers - This is going to cause issues with some parsers that expect headers

As I mentioned above, I would recommend the chain being a `repeated string` of PEM-encoded certificates. PKCS#12 is not widely supported.

Reply:
> Base64-encoded DER without the headers - This is going to cause issues with some parsers that expect headers

What ecosystems don't have easy ways to parse DER keys (especially PKIX SubjectPublicKeyInfo, which is what we pull out of X.509)?

- [Python cryptography](https://cryptography.io/en/3.4.6/hazmat/primitives/asymmetric/serialization.html#cryptography.hazmat.primitives.serialization.load_der_public_key)
- OpenSSL: use [d2i_PUBKEY_bio](https://www.openssl.org/docs/man3.0/man3/d2i_PUBKEY_bio.html) for PKIX and [d2i_RSAPublicKey_bio](https://www.openssl.org/docs/man1.1.1/man3/d2i_RSAPublicKey_bio.html). It's clunky, but parsing PEM isn't any better.
- Go: [ParsePKIXPublicKey](https://pkg.go.dev/crypto/x509#ParsePKIXPublicKey), [ParsePKCS1PublicKey](https://pkg.go.dev/crypto/x509#ParsePKCS1PublicKey); to parse PEM, [sigstore/sigstore just decodes it and then calls these](https://github.com/sigstore/sigstore/blob/56bd9cd0e7e712af8e22c041c471c6176f268bf1/pkg/cryptoutils/publickey.go#L53)
- Rust:
  - in the [RustCrypto world](https://github.com/RustCrypto/formats), you'd use [spki::SubjectPublicKeyInfo::try_from::<&[u8]>](https://docs.rs/spki/0.6.0/spki/struct.SubjectPublicKeyInfo.html#impl-TryFrom%3C%26%27a%20%5Bu8%5D%3E); same for [PKCS#1](https://docs.rs/pkcs1/latest/pkcs1/struct.RsaPublicKey.html#impl-TryFrom%3C%26%27a%20%5Bu8%5D%3E). If you want PEM, you have to do it yourself.
  - sigstore-rs uses x509_parser, which supports [FromDer](https://docs.rs/x509-parser/latest/x509_parser/prelude/trait.FromDer.html) for [SPKI](https://docs.rs/x509-parser/latest/x509_parser/x509/struct.SubjectPublicKeyInfo.html)
- Java: sigstore-java uses BouncyCastle; in BouncyCastle, you [construct a SubjectPublicKeyInfo from an ASN.1 object](https://www.bouncycastle.org/docs/docs1.8on/index.html)

PEM just feels like a really weird fit here—the point of PEM is that sending bytes over email is hard. But if we have a good way to send around bytes, and can document exactly what format they're going to be in, we don't need it. It's also strictly more bandwidth than base64.

Reply:
I was just meaning that you have to be aware it's base64 encoded DER and not a headerless PEM encoding.

I also want to avoid switching encoding around between different parts of Sigstore. We're using PEM in fulcio and rekor already, though rekor is currently doing some double encoding that's not ideal.

Reply:
I opted for PEM encoding here to be consistent with other parts of Sigstore as @haydentherapper mentioned. If we think that this spec should take the lead on setting the new formats as @znewman01 proposes, `repeated bytes` is what I would prefer, where the `bytes` is the DER encoded cert.

Reply:
I'd kinda prefer to stick with PEM encodings (`repeated string`), just since I assume these will stored as JSON typically and the DER encoding will be base64 encoded anyways

Reply:
>  since I assume these will stored as JSON typically and the DER encoding will be base64 encoded anyways

Yeah, but Protobuf makes that transparent to end users. I guess the question is: are other clients going to use these protos? I think we should be encouraging that, in which case I'd vote DER `repeated bytes`. If not, maybe PEM (as `repeated string`) does make sense because it's more self-documenting. I still think I prefer DER-bytes because PEM's often an extra dependency in a lot of languages, and it's one more step in parsing.

I can definitely be overruled here.

Reply:
> I guess the question is: are other clients going to use these protos?

I think this is an important question to answer, as there are some other fields too that are bytes, but returned either base64 or hex-encoded from Rekor. And I think we should be consistent to have either bytes or base64/hex encoded content in the protos. Like the `log id` as discussed here: https://github.com/sigstore/cosign/pull/2204#discussion_r968367700
`log id` now is the sha256 digest of the public key, but may be replaced with an OID (which is DER encoded IIRC).

If some fields are base64/PEM/hex I think all fields should follow for consistency.

Reply:
> log id now is the sha256 digest of the public key, but may be replaced with an OID 

Per [RFC 9162](https://www.rfc-editor.org/rfc/rfc9162.html#name-log-id), the log ID itself is an opaque 2-to-127-byte vector. It gets passed around DER-encoded in various APIs.

To my mind, that means we should make the type `bytes` in the proto. There's no PEM header for it AFAICT.

> And I think we should be consistent to have either bytes or base64/hex encoded content in the protos.
> [...]
> If some fields are base64/PEM/hex I think all fields should follow for consistency.

First, let me say that I think we should avoid hex-encoding anything.

I guess this is the core point of disagreement. I don't personally care at all what the encoding looks like, just what the parsed data structure looks like to client libraries after parsing; it should be as easy as possible to deal with at that point. If we have a `cosign inspect` tool, we could even store it on disk in a binary format. That's perhaps needlessly user-hostile though 😄 

IMO consistency in the encoding of values that are fundamentally different types is less important than representing and documenting each type well on its own. There's a totally reasonable argument in the opposite direction, though: we should err on the side of pragmatism, since JSON is the first thing many consumers will see, and enabling them to "parse by guessing" would be nice.

Reply:
> I guess this is the core point of disagreement.

I don't know, I think we may be talking past each other. And to be clear, I'm not proposing hex-encoding :) Just stating the fact that Rekor does return a lot of data hex encoded (log id, root hash, hashes in inclusion proof etc).

What I meant is what is the purpose of the proto files. Define the actual data types and structures, or also how to encode them? As if we rely on `bytes` in the proto for _binary_ fields, such as log ID, hashes etc. we only define the data types, in the most natural form. The encoding is left for the serialization, like base64 for JSON. 

If we on the other hand use `string` in the proto with some encoding, we go beyond just defining the data type. And here I think we may miss the purpose of using protos, as protobuf defines a mapping to e.g. JSON for sharing data between systems. @haydentherapper mentioned the (possibly) unexpected base64 encoding of `bytes`, I don't think this is any problematic. 
And from a developer perspective having the protofile and from that be able to generate data types/structures that are natively easy to use is beneficial. All the encoding/decoding is done automatically, and we spare the developer from doing explicit encoding/decoding prior/after marshalling/unmarshalling the JSON.

But with that said, I think: `we should err on the side of pragmatism, since JSON is the first thing many consumers will see` is reasonable too. As _if_ we rely on `bytes` in the protofiles, the JSON encoding will look different from the responses from Rekor, and that may confuse developers.

With proper documentation we can go either way, but my personable preference would be to rely on `bytes`, but I'm ok with using `strings` containing the values as returned by e.g. Rekor, or PEM encoded certificates.

Reply:
> What I meant is what is the purpose of the proto files. Define the actual data types and structures, or also how to encode them?

I think that's a good way to frame the problem.

I come down on the side of "use protos, make them as idiomatic as possible once parsed, the encoding is an accident." But I see why you'd want to treat the JSON encoding as the canonical format.

I think we should resolve this question *across* all Sigstore APIs/interfaces, as that's a potential source of inconsistency (though out-of-scope for this PR 😄)

Reply:
> I think we should resolve this question across all Sigstore APIs/interfaces, as that's a potential source of inconsistency

Totally agree to that!

Reply:
> FYI I added this as an agenda item to the Sigstore community meeting this coming Tuesday

Nice I should try and join too then.

Reply:
One thing to consider is that PEM headers do serve some purpose for keys. For example, differentiating between PKCS1 and PKIX encodings for RSA vs ECDSA.

I don't have a strong opinion on bytes vs strings for IDs. I'd probably say bytes to avoid hex encoding.

Reply:
I don't think this is a concern (PKCS1 vs PKIX)? As we won't encode the public key, only a hint. Only certificates will be encoded, and my understanding is that the PEM headers do not add any value. 

Making the hint `bytes` is ok with me. But looking at the DSSE spec again, they are relying on `string`, so maybe we should align with that? In DSSE it's called `keyid` and here it's currently named `hint`. I'm in favour of aligning with DSSE: `string keyid`, as there is a risk of confusion otherwise?

Reply:
w/r/t encoding: see https://github.com/sigstore/community/discussions/136

Reply:
Can we resolve this conversation and wait for the outcome of the linked discussion? As the discussion is mostly about encoding.